File: src/scenesplit/embeddings.py
"""Semantic embedding computation for video frames."""

from dataclasses import dataclass
from typing import Sequence

import numpy as np
import torch
import torchvision.models as models
import torchvision.transforms as transforms
from numpy.typing import NDArray
from PIL import Image

from scenesplit.constants import QualityPreset
from scenesplit.video import Frame


@dataclass(frozen=True)
class EmbeddedFrame:
    """A frame with its computed embedding vector."""

    frame: Frame
    embedding: NDArray[np.float32]

    @property
    def index(self) -> int:
        """Original frame index."""
        return self.frame.index

    @property
    def timestamp_seconds(self) -> float:
        """Frame timestamp in seconds."""
        return self.frame.timestamp_seconds


class EmbeddingModel:
    """Compute semantic embeddings for video frames using a pretrained vision model.

    Uses ResNet50 pretrained on ImageNet to extract feature vectors.
    All computation is done locally without network access.
    """

    def __init__(self, quality: QualityPreset = QualityPreset.BALANCED) -> None:
        """Initialize the embedding model.

        Args:
            quality: Quality preset affecting batch size and image scaling.
        """
        self.quality = quality
        self.device = self._get_device()

        # Load pretrained ResNet50 and remove the final classification layer
        # to get 2048-dimensional feature vectors
        self._model = models.resnet50(weights=models.ResNet50_Weights.IMAGENET1K_V2)
        self._model = torch.nn.Sequential(*list(self._model.children())[:-1])
        self._model = self._model.to(self.device)
        self._model.eval()

        # Preprocessing transform for ImageNet-pretrained models
        self._transform = transforms.Compose([
            transforms.Resize(256),
            transforms.CenterCrop(224),
            transforms.ToTensor(),
            transforms.Normalize(
                mean=[0.485, 0.456, 0.406],
                std=[0.229, 0.224, 0.225],
            ),
        ])

    def _get_device(self) -> torch.device:
        """Get the best available compute device."""
        if torch.cuda.is_available():
            return torch.device("cuda")
        elif torch.backends.mps.is_available():
            return torch.device("mps")
        return torch.device("cpu")

    def _preprocess_frame(self, frame: Frame) -> torch.Tensor:
        """Preprocess a frame for the embedding model.

        Args:
            frame: Frame to preprocess.

        Returns:
            Preprocessed tensor ready for the model.
        """
        # OpenCV uses BGR, convert to RGB
        rgb_data = frame.data[:, :, ::-1]

        # Apply resize factor based on quality
        resize_factor = self.quality.image_resize_factor
        if resize_factor < 1.0:
            h, w = rgb_data.shape[:2]
            new_h, new_w = int(h * resize_factor), int(w * resize_factor)
            import cv2
            rgb_data = cv2.resize(rgb_data, (new_w, new_h))

        # Convert to PIL Image and apply transforms
        image = Image.fromarray(rgb_data)
        return self._transform(image)

    def compute_embedding(self, frame: Frame) -> EmbeddedFrame:
        """Compute the semantic embedding for a single frame.

        Args:
            frame: Frame to embed.

        Returns:
            EmbeddedFrame with the computed embedding.
        """
        tensor = self._preprocess_frame(frame).unsqueeze(0).to(self.device)

        with torch.no_grad():
            embedding = self._model(tensor)

        # Flatten and normalize
        embedding = embedding.squeeze().cpu().numpy()
        embedding = embedding / np.linalg.norm(embedding)

        return EmbeddedFrame(frame=frame, embedding=embedding.astype(np.float32))

    def compute_embeddings_batch(
        self,
        frames: Sequence[Frame],
        progress_callback: callable | None = None,
    ) -> list[EmbeddedFrame]:
        """Compute embeddings for a batch of frames.

        Args:
            frames: Sequence of frames to embed.
            progress_callback: Optional callback(current, total) for progress.

        Returns:
            List of EmbeddedFrames with computed embeddings.
        """
        if not frames:
            return []

        batch_size = self.quality.embedding_batch_size
        results: list[EmbeddedFrame] = []

        for i in range(0, len(frames), batch_size):
            batch_frames = frames[i:i + batch_size]
            batch_tensors = torch.stack([
                self._preprocess_frame(f) for f in batch_frames
            ]).to(self.device)

            with torch.no_grad():
                embeddings = self._model(batch_tensors)

            # Process each embedding
            for j, frame in enumerate(batch_frames):
                embedding = embeddings[j].squeeze().cpu().numpy()
                embedding = embedding / np.linalg.norm(embedding)
                results.append(EmbeddedFrame(
                    frame=frame,
                    embedding=embedding.astype(np.float32),
                ))

            if progress_callback is not None:
                progress_callback(len(results), len(frames))

        return results


def cosine_similarity(a: NDArray[np.float32], b: NDArray[np.float32]) -> float:
    """Compute cosine similarity between two normalized embedding vectors.

    Args:
        a: First embedding vector (normalized).
        b: Second embedding vector (normalized).

    Returns:
        Cosine similarity in range [-1, 1].
    """
    return float(np.dot(a, b))

File: src/embeddings.rs
//! Semantic embedding computation for video frames.

use std::path::Path;

use ndarray::{s, Array4};
use ort::session::{builder::GraphOptimizationLevel, Session};

use crate::config::QualityPreset;
use crate::error::{Error, Result};
use crate::video::Frame;

/// A frame with its computed embedding vector.
#[derive(Debug, Clone)]
pub struct EmbeddedFrame {
    pub frame: Frame,
    pub embedding: Vec<f32>,
}

impl EmbeddedFrame {
    /// Original frame index.
    pub fn index(&self) -> usize {
        self.frame.index
    }

    /// Frame timestamp in seconds.
    pub fn timestamp_seconds(&self) -> f64 {
        self.frame.timestamp_seconds
    }
}

/// Compute semantic embeddings for video frames using ONNX Runtime.
pub struct EmbeddingModel {
    session: Session,
    quality: QualityPreset,
}

impl EmbeddingModel {
    /// Create a new embedding model.
    ///
    /// # Arguments
    ///
    /// * `model_path` - Path to the ONNX model file (ResNet50 or similar).
    /// * `quality` - Quality preset affecting image preprocessing.
    pub fn new<P: AsRef<Path>>(model_path: P, quality: QualityPreset) -> Result<Self> {
        let session = Session::builder()?
            .with_optimization_level(GraphOptimizationLevel::Level3)?
            .with_intra_threads(4)?
            .commit_from_file(model_path)?;

        Ok(Self { session, quality })
    }

    /// Preprocess a frame for the embedding model.
    ///
    /// Resizes to 224x224, normalizes with ImageNet mean/std.
    fn preprocess_frame(&self, frame: &Frame) -> Result<Array4<f32>> {
        let resize_factor = self.quality.image_resize_factor();

        // Calculate target size after optional resize
        let (src_width, src_height) = if resize_factor < 1.0 {
            (
                (frame.width as f32 * resize_factor) as u32,
                (frame.height as f32 * resize_factor) as u32,
            )
        } else {
            (frame.width, frame.height)
        };

        // Create image from raw RGB data
        let img = image::RgbImage::from_raw(frame.width, frame.height, frame.data.clone())
            .ok_or_else(|| Error::Embedding("Failed to create image from frame data".into()))?;

        // Resize if needed
        let img = if resize_factor < 1.0 {
            image::imageops::resize(
                &img,
                src_width,
                src_height,
                image::imageops::FilterType::Triangle,
            )
        } else {
            img
        };

        // Resize to 224x224 for the model
        let img = image::imageops::resize(&img, 224, 224, image::imageops::FilterType::Triangle);

        // Convert to NCHW format with normalization
        // ImageNet mean: [0.485, 0.456, 0.406], std: [0.229, 0.224, 0.225]
        let mean = [0.485f32, 0.456, 0.406];
        let std = [0.229f32, 0.224, 0.225];

        let mut tensor = Array4::<f32>::zeros((1, 3, 224, 224));

        for y in 0..224 {
            for x in 0..224 {
                let pixel = img.get_pixel(x as u32, y as u32);
                for c in 0..3 {
                    let value = pixel[c] as f32 / 255.0;
                    let normalized = (value - mean[c]) / std[c];
                    tensor[[0, c, y, x]] = normalized;
                }
            }
        }

        Ok(tensor)
    }

    /// Compute the embedding for a single frame.
    pub fn compute_embedding(&mut self, frame: &Frame) -> Result<EmbeddedFrame> {
        let input = self.preprocess_frame(frame)?;
        let input_value = ort::value::Tensor::from_array(input)?;
        let outputs = self.session.run(ort::inputs![input_value])?;

        // Get the output tensor - new API returns (shape, data) tuple
        let (_, data) = outputs[0]
            .try_extract_tensor::<f32>()
            .map_err(|e| Error::Embedding(format!("Failed to extract embedding: {}", e)))?;

        // Flatten and normalize
        let flat: Vec<f32> = data.iter().cloned().collect();
        let embedding = normalize_vector(&flat);

        Ok(EmbeddedFrame {
            frame: frame.clone(),
            embedding,
        })
    }

    /// Compute embeddings for a batch of frames.
    pub fn compute_embeddings_batch<F>(
        &mut self,
        frames: &[Frame],
        mut progress_callback: Option<F>,
    ) -> Result<Vec<EmbeddedFrame>>
    where
        F: FnMut(usize, usize),
    {
        if frames.is_empty() {
            return Ok(Vec::new());
        }

        let batch_size = self.quality.embedding_batch_size();
        let mut results = Vec::with_capacity(frames.len());

        for (batch_idx, chunk) in frames.chunks(batch_size).enumerate() {
            // Process batch
            let mut batch_tensor = Array4::<f32>::zeros((chunk.len(), 3, 224, 224));

            for (i, frame) in chunk.iter().enumerate() {
                let preprocessed = self.preprocess_frame(frame)?;
                batch_tensor
                    .slice_mut(s![i, .., .., ..])
                    .assign(&preprocessed.slice(s![0, .., .., ..]));
            }

            let batch_value = ort::value::Tensor::from_array(batch_tensor)?;
            let outputs = self.session.run(ort::inputs![batch_value])?;

            let (shape, data) = outputs[0]
                .try_extract_tensor::<f32>()
                .map_err(|e| Error::Embedding(format!("Failed to extract embeddings: {}", e)))?;

            // Calculate embedding size from shape (batch_size, embedding_dim, ...)
            let embedding_size = shape.iter().skip(1).product::<i64>() as usize;

            // Extract individual embeddings from batch output
            for (i, frame) in chunk.iter().enumerate() {
                let start = i * embedding_size;
                let end = start + embedding_size;
                let flat: Vec<f32> = data
                    .iter()
                    .skip(start)
                    .take(embedding_size)
                    .cloned()
                    .collect();
                let embedding = normalize_vector(&flat);

                results.push(EmbeddedFrame {
                    frame: frame.clone(),
                    embedding,
                });
            }

            if let Some(ref mut cb) = progress_callback {
                cb(results.len(), frames.len());
            }
        }

        Ok(results)
    }
}

/// Normalize a vector to unit length.
fn normalize_vector(v: &[f32]) -> Vec<f32> {
    let norm: f32 = v.iter().map(|x| x * x).sum::<f32>().sqrt();
    if norm > 0.0 {
        v.iter().map(|x| x / norm).collect()
    } else {
        v.to_vec()
    }
}

/// Compute cosine similarity between two normalized embedding vectors.
pub fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
    a.iter().zip(b.iter()).map(|(x, y)| x * y).sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cosine_similarity_identical() {
        let v = vec![1.0, 0.0, 0.0];
        assert!((cosine_similarity(&v, &v) - 1.0).abs() < 1e-6);
    }

    #[test]
    fn test_cosine_similarity_orthogonal() {
        let v1 = vec![1.0, 0.0, 0.0];
        let v2 = vec![0.0, 1.0, 0.0];
        assert!(cosine_similarity(&v1, &v2).abs() < 1e-6);
    }

    #[test]
    fn test_cosine_similarity_opposite() {
        let v1 = vec![1.0, 0.0, 0.0];
        let v2 = vec![-1.0, 0.0, 0.0];
        assert!((cosine_similarity(&v1, &v2) + 1.0).abs() < 1e-6);
    }

    #[test]
    fn test_normalize_vector() {
        let v = vec![3.0, 4.0];
        let normalized = normalize_vector(&v);
        let norm: f32 = normalized.iter().map(|x| x * x).sum::<f32>().sqrt();
        assert!((norm - 1.0).abs() < 1e-6);
    }
}
